### ISSUE TITLE
Improve handling of files on network shares

### ISSUE THREAD

**Issue by oxan** [MPExtended/MPExtended#48]
```
[23:25:22] <DieBagger> http://forums.asp.net/t/923580.aspx
[23:25:39] <DieBagger> do you use the same user for the service as your local login?
[23:26:44] <DieBagger> btw oxan, did you know about impersonation? http://www.codeproject.com/KB/cs/zetaimpersonator.aspx
[23:26:57] <oxan> no
[23:27:11] <oxan> looks interesting
[23:27:51] <Pservais> yes, same user as local login is used
[23:28:15] <DieBagger> I'm not sure how mapped network drives work
[23:28:35] <DieBagger> the article seems to suggest that they exist on a per-session basis
[23:28:59] <oxan> yeah afaik they're on a per-session basis with an option to make them per-user (which almost all people use)
[23:29:54] <DieBagger> When I add a mapped network drive I see no option to do such a thing
[23:30:46] <oxan> maybe you can only do that on the command line then
[23:30:56] <oxan> net use X:\ \\server\bla /persisent:yes 
[23:32:04] <DieBagger> hm, it seems persistent is the same as "reconnect at startup"
[23:32:23] <DieBagger> but the question is if this is visible for services running under the user
[23:32:27] <DieBagger> it would seem not
[23:33:26] <DieBagger> but maybe user impersonation would fix our problem
[23:34:01] <DieBagger> it would also make it easier (no need to change the services' user)
[23:34:23] <DieBagger> btw
[23:34:39] <DieBagger> we should add a warning for files on nas systems
[23:34:40] <oxan> yeah we should try if impersonation helps us
[23:34:44] <oxan> yep
[23:34:52] <DieBagger> so e.g. if a file is requested and access is denied
[23:34:58] <DieBagger> we check if it starts with "\\\\"
[23:35:30] <DieBagger> if yes, we log a warning "we cant access the file and it seems to be located on a network share, make sure you run under a user with access"
[23:35:33] <DieBagger> or smth like that
[23:35:39] <oxan> i think that we can expect more problems with mapped network drives then with UNC paths
[23:35:41] <oxan> but yes, good idea
```

We should either handle it in MediaSource or in the GetFileInfo and friends methods. It needs a bit of redesign to avoid copying the impersonation code all over the place I'm afraid.

**Comment by DieBagger** [comment]
What about ffmpeg/vlc? Will they also run in the impersonated context?

**Comment by oxan** [comment]
I think i'll let ffmpeg/vlc read from a named pipe (like we do for TV), so that we can run them in the normal context. Should check if that gives regressions though. That way I can abstract the impersonation to the file handling in MAS. Otherwise I should add it in a lot of places which creates the chance that we forget a place.

**Comment by oxan** [comment]
- We can't use the linked code because it's under CPOL which isn't compatible with GPL (it isn't very hard though: http://support.microsoft.com/default.aspx?scid=kb;en-us;Q306158)
- For MAS and streaming we can easily implement it in MediaAccessService.RetrieveStream(), which is used by all other file access methods (IsLocalFile returns false because Exists = false)
- For recordings we can put it in MediaSource; which should be used by all streaming service components (IsLocalFile should check for File.Exists() and GetStream() should do the impersonation). 
- So the impersonation implementation should go into the general library

**Comment by oxan** [comment]
And implemented for recordings in e20be08.
